posts/2019-07-04-Rust-1.36.0.md

```diff
@@ -0,0 +1,164 @@
+---
+layout: post
+title: "Announcing Rust 1.36.0"
+author: The Rust Release Team
+release: true
+---
+
+The Rust team is happy to announce a new version of Rust, 1.36.0. Rust is a programming language that is empowering everyone to build reliable and efficient software.
+
+If you have a previous version of Rust installed via rustup, getting Rust 1.36.0 is as easy as:
+
+```console
+$ rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`][install] from the appropriate page on our website, and check out the [detailed release notes for 1.36.0][notes] on GitHub.
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1360-2019-07-04
+
+## What's in 1.36.0 stable
+
+This release brings many changes, including the stabilization of the [`Future`] trait, the [`alloc`][alloc-crate] crate, the [`MaybeUninit<T>`] type, [NLL for Rust 2015][felix-blog], a new `HashMap<K, V>` implementation, and [`--offline`] support in Cargo. Read on for a few highlights, or see the [detailed release notes][notes] for additional information.
+
+### The [`Future`] is here!
+
+[`Future`]: https://doc.rust-lang.org/std/future/trait.Future.html
+[pr-future]: https://github.com/rust-lang/rust/pull/59739
+
+In Rust 1.36.0 the long awaited [`Future`] trait has been [stabilized][pr-future]!
+
+With this stabilization, we hope to give important crates, libraries, and the ecosystem time to prepare for `async` / `.await`, which we'll tell you more about in the future.
+
+### The [`alloc`][alloc-crate] crate is stable
+
+[alloc-crate]: https://doc.rust-lang.org/alloc/index.html
+
+Before 1.36.0, the standard library consisted of the crates `std`, `core`, and `proc_macro`. The `core` crate provided core functionality such as `Iterator` and `Copy` and could be used in `#![no_std]` environments since it did not impose any requirements. Meanwhile, the `std` crate provided types like `Box<T>` and OS functionality but required a global allocator and other OS capabilities in return.
+
+Starting with Rust 1.36.0, the parts of `std` that depend on a global allocator, e.g. `Vec<T>`, are now defined in the `alloc` crate. The `std` crate then re-exports these parts. While `#![no_std]` binaries using `alloc` still require nightly Rust, `#![no_std]` library crates can use the `alloc` crate in stable Rust, and normal binaries, without `#![no_std]`, can depend on such library crates. We hope this will facilitate the development of a `#![no_std]` compatible ecosystem of libraries prior to stabilizing support for `#![no_std]` binaries using `alloc`.
```

Comment:
> While `#![no_std]` binaries using `alloc` still require nightly Rust,

I think this is either wrong or misleading. As far as I know `#![no_std]` binaries in general, independently of the `alloc` crate, require Nightly on many platforms because of ``error: requires `start` lang_item``. (But for some micro-controller / embedded targets the `start` lang items is not involved, and full “binaries” can be built on stable with a custom linker script.)

Consider rewriting to something like:

> While `#![no_std]` binaries still require nightly Rust,

Reply:
Heh; this is what the blog post said at the start but I changed it after @jonas-schievink pointed out that you can use `#![no_main]` on stable and stuff. (https://github.com/jonas-schievink/rubble/blob/master/rubble-demo/src/main.rs)

Reply:
Oh, never mind. I was gonna write that it’s still not liballoc that adds a Nigthly-only requirement but that’s not true: it requires `#[alloc_error_handler]` which is not stable. (I thought it was.) https://github.com/rust-lang/rust/issues/51540

So yeah, “`#![no_std]` binaries using `alloc` still require nightly Rust” is accurate in 1.36.

Reply:
Alright; thanks for double checking! =)